precovery/brute_force.py

```diff
@@ -0,0 +1,387 @@
+        # Query tree using observed RA, Dec
+        d, i = tree.query(
+            coords_latlon,
+            k=k,
+            return_distance=True,
+            dualtree=True,
+            breadth_first=False,
+            sort_results=False,
+        )
+
+        # Select all observations with distance smaller or equal
+        # to the maximum given distance
+        mask = np.where(d <= eps_rad)
+
+        if len(d[mask]) > 0:
+            orbit_ids_associated.append(orbit_ids[i[mask]])
+            obs_ids_associated.append(obs_ids[mask[0]])
+            obs_times_associated.append(obs_times[mask[0]])
+            distances.append(d[mask].reshape(-1, 1))
+
+            residuals_visit, stats_visit = calcResiduals(
+                coords[mask[0]],
+                coords_predicted[i[mask]],
+                sigmas_actual=coords_sigma[mask[0]],
+                include_probabilistic=True,
+            )
+            residuals.append(residuals_visit)
+            stats.append(np.vstack(stats_visit).T)
+
+    if len(distances) > 0:
+        distances = np.degrees(np.vstack(distances))
+        orbit_ids_associated = np.vstack(orbit_ids_associated)
+        obs_ids_associated = np.vstack(obs_ids_associated)
+        obs_times_associated = np.vstack(obs_times_associated)
+        residuals = np.vstack(residuals)
+        stats = np.vstack(stats)
+
+        attributions = {
+            "orbit_id": orbit_ids_associated[:, 0],
+            "obs_id": obs_ids_associated[:, 0],
+            "mjd_utc": obs_times_associated[:, 0],
+            "distance": distances[:, 0],
+            "residual_ra_arcsec": residuals[:, 0] * 3600,
+            "residual_dec_arcsec": residuals[:, 1] * 3600,
+            "chi2": stats[:, 0],
+        }
+        if include_probabilistic:
+            attributions["probability"] = stats[:, 1]
+            attributions["mahalanobis_distance"] = stats[:, 2]
+
+        attributions = pd.DataFrame(attributions)
+
+    else:
+        columns = [
+            "orbit_id",
+            "obs_id",
+            "mjd_utc",
+            "distance",
+            "residual_ra_arcsec",
+            "residual_dec_arcsec",
+            "chi2",
+        ]
+        if include_probabilistic:
+            columns += ["probability", "mahalanobis_distance"]
+
+        attributions = pd.DataFrame(columns=columns)
+
+    return attributions
+
+
+def attributeObservations(
+    orbits,
+    mjd_start: float,
+    mjd_end: float,
+    precovery_db: PrecoveryDatabase,
+    eps=5 / 3600,
+    include_probabilistic=True,
+    backend="PYOORB",
+    backend_kwargs={},
+    orbits_chunk_size=10,
+    observations_chunk_size=100000,
+    num_jobs=1,
+    parallel_backend="mp",
+):
+    logger.info("Running observation attribution...")
+    time_start = time.time()
+
+    num_orbits = len(orbits)
+
+    attribution_dfs = []
+
+    # prepare ephemeris and observation dictionaries
+    ephemeris, observations = get_observations_and_ephemerides_for_orbits(
```

Comment:
I may be completely misunderstanding this, so I apologize if I'm off.
Can this function be broken up into separate functions? The process seems to be:

1. orbits -> ephemerides
2. ephemerides -> all unique intersecting frames + neighbors
3. frames -> get all observations 

The reason being that I believe most of these are re-usable and consistent and may simply require parameters to adjust results based on brute-force vs. nelson-force, etc. Here is what I am thinking this looks like alternatively,

```python
frame_times = get_frame_times_in_range(mjd_start, mjd_end)
ephemerides = ephemerides_from_orbits(orbits, epochs, obscode=obscode)
frames_to_search = intersecting_frames(ephemerides, neighbors=True) # Default is False
observations = precovery_db.extract_observations_by_frames(filtered_frames)

```

I think each of these steps is usable in multiple cases, especially with using optional parameters.
This makes each individual piece more testable and gives back simpler return values.